### ISSUE TITLE
`select_all` function errors when retrieving certain tables from a base

### ISSUE THREAD

**Issue by ernestguevarra** [ecohealthalliance/airtabler#49]
**Describe the bug**
When using the `select_all()` function to retrieve all records/rows of a specific table from a base, error arises for some tables. The error message is:

```
Error in `.rowNamesDF<-`(x, value = value) : 
  duplicate 'row.names' are not allowed
In addition: Warning message:
non-unique values when setting 'row.names': ‘1’, ‘10’, ‘100’, ‘11’, ‘12’, ‘13’, ‘14’, ‘15’, ‘16’,
‘17’, ‘18’, ‘19’, ‘2’, ‘20’, ‘21’, ‘22’, ‘23’, ‘24’, ‘25’, ‘26’, ‘27’, ‘28’, ‘29’, ‘3’, ‘30’,
‘31’, ‘32’, ‘33’, ‘34’, ‘35’, ‘36’, ‘37’, ‘38’, ‘39’, ‘4’, ‘40’, ‘41’, ‘42’, ‘43’, ‘44’, ‘45’,
‘46’, ‘47’, ‘48’, ‘49’, ‘5’, ‘50’, ‘51’, ‘52’, ‘53’, ‘54’, ‘55’, ‘56’, ‘57’, ‘58’, ‘59’, ‘6’,
‘60’, ‘61’, ‘62’, ‘63’, ‘64’, ‘65’, ‘66’, ‘67’, ‘68’, ‘69’, ‘7’, ‘70’, ‘71’, ‘72’, ‘73’, ‘74’,
‘75’, ‘76’, ‘77’, ‘78’, ‘79’, ‘8’, ‘80’, ‘81’, ‘82’, ‘83’, ‘84’, ‘85’, ‘86’, ‘87’, ‘88’, ‘89’,
‘9’, ‘90’, ‘91’, ‘92’, ‘93’, ‘94’, ‘95’, ‘9 [... truncated] 
```

This error message seems to indicate that the issue is arising from when the records from each page retrieved are concatenated into a single data.frame.

**To Reproduce**

I have added @collinschwantes as a creator in the dev version of the airtable base from which I am getting this issue. The following code should reproduce the issue

```R
## Connect with airtable base and select tables to work with
acdc_tables <- airtable(
  base = Sys.getenv("AIRTABLE_ACDC_DEV"),
  tables = c("Researchers", "Institutions", "Activities", "Topics", "AU Countries")
)

## Get ACDC researchers table - no issues using either `select()` or `select_all()`

### Use `select()`
acdc_researchers <- acdc_tables$Researchers$select()

### Use `select_all()`
acdc_researchers <- acdc_tables$Researchers$select_all()

## Get ACDC activities table - no issues using `select()` but error arises when using `select_all()`

### Use `select()`
acdc_activities <- acdc_tables$Activities$select()

### Use `select_all()`
acdc_activties <- acdc_tables$Activities$select_all()
```

**Expected behavior**
Expected behaviour is that `select_all()` would provide all the records for the **Activities** table in a single data.frame in the same way as it does for the **Researchers** table.

**Session info (please complete the following information):**

I don't think the session info will be relevant here as I think I have narrowed down the issue to the unexported function in the `airtabler` package called `.bind_df()` which is shown below:

```R
.bind_df <- function (x) 
{
    if (length(unique(lengths(x))) != 1) {
        col_names <- unique(unlist(lapply(x, names)))
        col_missing <- lapply(x, function(x) setdiff(col_names, 
            names(x)))
        x <- lapply(seq_along(x), function(i) {
            ret <- x[[i]]
            for (col in col_missing[[i]]) {
                ret[[col]] <- NA
            }
            ret
        })
    }
    do.call(rbind, x)
}
```

See notes below on further troubleshooting done related to the `.bind_df()` function.

**Additional context**

Noting that the error message seems to indicate an issue with how the records from multiple pages are being concatenated, I looked into how the `select_all()` function is formulated. The function is below:

```R
$select
function (record_id = NULL, fields = NULL, filterByFormula = NULL, 
    maxRecord = NULL, sort = NULL, view = NULL, pageSize = NULL, 
    offset = NULL, combined_result = TRUE) 
{
    air_select(base, table_name, record_id, fields, filterByFormula, 
        maxRecord, sort, view, pageSize, offset, combined_result)
}
<bytecode: 0x55c80fa8fe70>
<environment: 0x55c80f9ee0e0>

$select_all
function (record_id = NULL, fields = NULL, filterByFormula = NULL, 
    maxRecord = NULL, sort = NULL, view = NULL, pageSize = NULL) 
{
    ret_all <- list()
    ret_offset = NULL
    while ({
        ret <- air_select(base, table_name, record_id, fields, 
            filterByFormula, maxRecord, sort, view, pageSize, 
            ret_offset, combined_result = TRUE)
        ret_offset <- get_offset(ret)
        ret_all <- c(ret_all, list(ret))
        !is.null(ret_offset)
    }) {
    }
    if (length(ret_all) == 0) {
        return(list())
    }
    .bind_df(ret_all)
}
<bytecode: 0x55c80fa8f460>
<environment: 0x55c80f9ee0e0>
```

The `select_all()` function iterates the `select()` function to each page of 100 records of a specified table based on a returned page offset value and with each of the returned page of 100 records concatenated into a list. The list is then simplified into a single data.frame using the `.bind_df()` function.

Based on the error message, the issue is most likely arising from when the `.bind_df()` function is applied to specific tables (it works in some tables and errors in other tables).

I reviewed the `.bind_df()` function (shown above) and it seems like the function first checks if the different elements in the list of data.frames of records have differing number of columns. If there are different number of columns, it checks for the column names in each element of the list against the complete column names, adds the missing column and then creates NA values for this column.

I think this is relevant to the Africa CDC base and its tables because certain tables have much parser information in certain fields/columns. When checking the Researchers table (which didn't error when using the `select_all()` function, the first 2 sets of 100 records had the same number of fields/columns. The third set/page of 100 records had one less columns while the fourth and last set/page of records had 2 less columns. On the other hand, the Activities table had differing columns/fields for each page/set of 100 records.

I think this is where the issue is happening. I am still trying to pinpoint exactly why the error message is complaining about non-unique row names. But it does seem likely that this is the function that is throwing the error.

I am wondering whether just using `dplyr::bind_rows()` to concatenate the list of data.frames will suffice? It seems the `.bind_df()` function was needed because the package author wanted to use the `do.call(rbind, x)` approach which requires for each of the data.frames to be concatenated to have the same column names. But from what I can remember, `dplyr::bind_rows()` already deals with this situation?

So, my instinct for now is write a function in my workflow that just uses a refactor of the airtabler function that uses `dplyr::bind_rows()` instead.

**Comment by ernestguevarra** [comment]
@collinschwantes just filing this issue for your attention and comment. I am unsure whether it is indeed an issue with the package functions or whether it manifests something wrong with how we structured the airtable base and its tables.

I detailed what I was experiencing and the results of my troubleshooting. I hope these are detailed enough.

If you think that this is an issue that a refactor in the package functions should address, I will be happy to make a pull request to address this. Otherwise, I will sort our airtable base and our tables accordingly so that we won't face this issue.
